src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+        if (hasTypeMismatchErrors(expectedParamTypes[i], actualParamTypes[i])) {
+          return true
+        }
+      }
+      // Check return type
+      return hasTypeMismatchErrors(actualType.returnType, expectedType.returnType)
+    case 'union':
+      const expectedSet = new Set(expectedType.types.map(formatTypeString))
+      // If not union type, expected set should contain actual type
+      if (actualType.kind !== 'union') {
+        return !expectedSet.has(formatTypeString(actualType))
+      }
+      // If both are union types, actual set should be a subset of expected set
+      const actualSet = new Set(actualType.types.map(formatTypeString))
+      for (const elem of actualSet) {
+        if (!expectedSet.has(elem)) {
+          return true
+        }
+      }
+      return false
+    default:
+      return true
+  }
+}
+
+/**
+ * Checks the types of the arguments of the given call expression.
+ * If number of arguments is different, add InvalidNumberOfArguments error to context and terminates early.
+ * Else, checks each argument against its expected type.
+ */
+function checkArgTypes(
+  node: es.CallExpression,
+  expectedTypes: Primitive[],
+  context: Context,
+  env: TypeEnvironment
+) {
+  const args = node.arguments
+  if (args.length !== expectedTypes.length) {
+    context.errors.push(new InvalidNumberOfArguments(node, expectedTypes.length, args.length))
+    return
+  }
+  for (let i = 0; i < expectedTypes.length; i++) {
+    const expectedType = expectedTypes[i]
+    const node = args[i]
+    const actualType = typeCheckAndReturnType(node, context, env)
+    checkForTypeMismatch(node, actualType, expectedType, context)
+  }
+}
+
+/**
+ * Converts type annotation node to its corresponding type representation in Source.
+ * If no type annotation exists, returns the "any" primitive type.
+ */
+function getAnnotatedType(
+  annotationNode: AnnotationTypeNode | TypeAliasDeclarationNode | undefined,
+  context: Context,
+  env: TypeEnvironment
+): Type {
+  if (!annotationNode) {
+    return tAny
+  }
+  const annotatedTypeNode = annotationNode.typeAnnotation
+  switch (annotatedTypeNode.type) {
+    case TSNodeType.TSFunctionType:
+      const fnTypeNode = annotatedTypeNode as FunctionTypeNode
+      const fnTypes = getParamTypes(fnTypeNode.parameters, context, env)
+      // Return type will always be last item in types array
+      fnTypes.push(getAnnotatedType(fnTypeNode.typeAnnotation, context, env))
+      return tFunc(...fnTypes)
+    case TSNodeType.TSUnionType:
+      const unionTypeNode = annotatedTypeNode as UnionTypeNode
+      const unionTypes = unionTypeNode.types.map(type =>
+        getPrimitiveType(annotationNode, type, context)
+      )
+      return mergeTypes(...unionTypes)
+    case TSNodeType.TSIntersectionType:
+      throw new TypeError(
+        annotationNode as unknown as es.Node,
+        'Intersection types are not allowed'
+      )
```

Comment:
A benefit of making use of discriminated unions is that we will find out that this case is impossible! Unless of course everything is already mistyped at this point...

Reply:
This is a special case where the type of annotationNode is not `es.Node` per se, but the shared attributes are enough such that the casting is safe. I'll try to find an alternative that doesn't involve typecasting though.